Enforce cms python types in configurations

Dr15Jones opened the pull request:
#### PR description:

- when assigning a parameter in python using an explicit type, require that existing value's type matches new type
- fixed problem where _ProxyParameter was not forwarding container methods to set value

#### PR validation:

Framework unit tests pass.

Dr15Jones committed:
Avoid unnecessary calls when assigning initial values from defaults

Dr15Jones committed:
Support container method forwarding in _ProxyParameter

Needed to handle cms.[vV]* classes properly. Modifying the tests for InputTag
search and replace uncovered this missing feature.

Dr15Jones committed:
Enforce assignment with explicit type matches previous type

In the python configuration, if a PSet parameter already has been assigned, then reassignment using an explicit type will only work if the types match.

Dr15Jones commented:
please test

Dr15Jones commented:
loosely associated with #41349

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41361/35208

- This PR adds an extra 76KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/ParameterSet/python/Config.py  modified in PR(s): #37996

cmsbuild commented:
A new Pull Request was created by  @Dr15Jones (Chris Jones) for master.

It involves the following packages:

- FWCore/ParameterSet (**core**)


@makortel, @smuzaffar, @Dr15Jones can you please review it and eventually sign? Thanks.
@makortel, @missirol, @wddgit this is something you requested to watch as well.
@perrotta, @dpiparo, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

cmsbuild commented:
-1

**Failed Tests**: UnitTests RelVals 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7082af/32007/summary.html
**COMMIT**: 19aab2b36500d0a118dcb316cfbf3afdd5fd3cab
**CMSSW**: CMSSW_13_1_X_2023-04-17-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/41361/32007/install.sh` to create a dev area with all the needed externals and cmssw changes.



The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - @makortel cms-sw/cmssw#41164

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7082af/32007/git-recent-commits.json
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7082af/32007/git-merge-result


## Unit Tests

I found errors in the following unit tests:

<pre>---> test runtestPhysicsToolsPatAlgos had ERRORS
</pre>

## RelVals

- [101.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7082af/32007/runTheMatrix-results/101.0_SingleElectronE120EHCAL)```101.0_SingleElectronE120EHCAL/step1_SingleElectronE120EHCAL.log```

Dr15Jones committed:
Can specify a default for `allowed` configuration parameter

In addition to specify the allowed types, one can now use the
`default` keyword to specify a starting value.

Dr15Jones committed:
Remove explicit use of type when resetting parameter in ECALHCAL.py

The type used was incorrect (a cms.string instead of a cms.InputTag).
Removing the type altogether should help long term maintenance.

Dr15Jones commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41361/35224

- This PR adds an extra 40KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/ParameterSet/python/Config.py  modified in PR(s): #37996

cmsbuild commented:
Pull request #41361 was updated. @smuzaffar, @civanch, @Dr15Jones, @swertz, @vlimant, @makortel, @micsucmed, @emanueleusai, @mdhildreth, @clacaputo, @syuvivida, @simonepigazzini, @pmandrik, @mandrenguyen, @rvenditti can you please check and sign again.

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7082af/32029/summary.html
**COMMIT**: 1b69814741b90c9f79b9705c6ccf1c1815deb372
**CMSSW**: CMSSW_13_1_X_2023-04-18-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/41361/32029/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found errors in the following unit tests:

<pre>---> test testFWCoreFrameworkWrongOptionsNumberOfStreamsType had ERRORS
---> test testFWCoreFrameworkWrongOptionsNumberOfThreadsType had ERRORS
---> test testFWCoreFrameworkWrongOptionssizeOfStackForThreadsInKBType had ERRORS
</pre>

## Comparison Summary



**Summary**:
 - You potentially removed 10 lines from the logs
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3459877
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3459849
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-04-18-1100+7082af/56578/triggerResults): no differences found

Dr15Jones committed:
Force check in C++ for wrong type unit test

Removing the parameter then adding it back avoids checking the
new values type in python.

makortel commented:
```
> ---> test testFWCoreFrameworkWrongOptionsNumberOfStreamsType had ERRORS
> ---> test testFWCoreFrameworkWrongOptionsNumberOfThreadsType had ERRORS
> ---> test testFWCoreFrameworkWrongOptionssizeOfStackForThreadsInKBType had ERRORS
```
I'd probably adjust these tests to use the `del process.options.foo; process.options.foo = <incorrect type:` to keep testing the C++ side check.

Dr15Jones commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7082af/32044/summary.html
**COMMIT**: 7c70dc4e70ef50ff11b168c6b7b5f01808dc9917
**CMSSW**: CMSSW_13_1_X_2023-04-19-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/41361/32044/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially removed 17 lines from the logs
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3459877
 - DQMHistoTests: Total failures: 0
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3459855
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-04-19-1100+7082af/56588/triggerResults): no differences found

emanueleusai commented:
+1

civanch commented:
+1

simonepigazzini commented:
+1

smuzaffar commented:
sounds reasonable to me @Dr15Jones . Please update the base branch to `CMSSW_13_1_DEVEL_X` and then we can ask @cms-sw/orp-l2  to merge it for devel IBs.

makortel commented on FWCore/ParameterSet/python/Config.py:
```diff
@@ -422,9 +422,9 @@ def __setattr__(self,name,value):
+        if name == 'options' and hasattr(self,name):
             value = self.__updateOptions(value)
-        if name == 'maxEvents':
+        if name == 'maxEvents' and hasattr(self,name):
```
The purpose and effect of these changes is not clear to me (even after reading their commit message). Could you elaborate?

Dr15Jones replied:
The constructor of `Process` has the line
```python
        self.options = Process.defaultOptions_()
        self.maxEvents = Process.defaultMaxEvents_()
```
Both of those trigger the `__setattr__` call. As the attributes have not been called yet, it is more efficient not to call the `__update*` methods and instead just set the attribute directly.

makortel replied:
What happens (or should happen) in the following cases?
```py
process = cms.Process("Test")
# 1
del process.options
process.options = cms.PSet(numberOfThreads = cms.untracked.uint32(5))
# 2
del process.options
process.options = cms.PSet(nonExistent = 6)
# 3
del process.options
process.options = dict(numberOfThreads = 7)
# 4
del process.options
process.options = dict(nonExistent = 8)
```

makortel commented on FWCore/ParameterSet/python/Types.py:
```diff
@@ -40,11 +40,27 @@ def __init__(self,type):
+    def _returnTypeWithValue(self, valueWithType):
+        if isinstance(valueWithType, type(self)):
+            return valueWithType
+        if isinstance(self.__type, type):
```
Where does the variable `type` come from? (or does it refer to the `type()` function?)

Dr15Jones replied:
`type` isn't actually a function, it is actually the base class for all `classes`. E.g.
```python3 
>>> import FWCore.ParameterSet.Config as cms
>>> type(cms.int32)
<class 'type'>
>>> if isinstance(cms.int32, type): print('True')
... 
True
```

makortel replied:
Ok

makortel commented on FWCore/ParameterSet/python/Types.py:
```diff
@@ -1919,6 +1960,8 @@ def testRequired(self):
             self.assertEqual(p1.dumpPython(),'cms.PSet(\n    foo = cms.PSet(\n        a = cms.int32(5)\n    ),\n    allowAnyLabel_=cms.required.PSetTemplate(\n        a = cms.required.int32\n    )\n)')
             self.assertEqual(p1.foo.a.value(), 5)
+            p1 = PSet(anInt = required.int32)
+            self.assertRaises(TypeError, lambda : setattr(p1,'anInt', uint32(2)))
```
These assertions could be also written as
```suggestion
            self.assertRaises(TypeError, setattr, p1,'anInt', uint32(2))
```
to avoid the layer of lambda, but I don't have strong feelings to either direction.

Dr15Jones replied:
changed

makortel commented on FWCore/ParameterSet/python/Types.py:
```diff
@@ -118,14 +153,13 @@ def _dumpPythonName():
+        self.__dict__['_default'] = None
         self.__dict__['__name__'] = self.dumpPython()
+        if len(kargs):
+            self.__dict__['_default'] = self._setValueWithType(kargs['default'])
```
IIUC, if the caller uses an incorrect keyword argument, the error is about the `kargs` not having an element `'default'` rather than about an unexpected keyword argument. How about something along
```py
def __init__(self, *args, default=None):
    # <snip>
    if default is not None:
        self.__dict__['_default'] = self._setValueWithType(default)
```
?

Dr15Jones replied:
changed

Dr15Jones committed:
Changed based on review comments

- explicitly used named parameter
- simplify a unit test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41361/35265

- This PR adds an extra 92KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/ParameterSet/python/Config.py  modified in PR(s): #37996

cmsbuild commented:
Pull request #41361 was updated. @smuzaffar, @civanch, @Dr15Jones, @swertz, @vlimant, @makortel, @micsucmed, @emanueleusai, @mdhildreth, @cmsbuild, @clacaputo, @syuvivida, @simonepigazzini, @pmandrik, @mandrenguyen, @rvenditti can you please check and sign again.

makortel commented on FWCore/ParameterSet/python/Types.py:
```diff
@@ -40,11 +40,27 @@ def __init__(self,type):
         if not _ParameterTypeBase.isTracked(self):
             v = untracked(v)
         self.__dict__["_ProxyParameter__value"] = v
+    def _returnTypeWithValue(self, valueWithType):
```
Does the "type with value" and "value with type" have any difference? I see "value with type" being used more, and am wondering if `_returnValueWithType` would make more sense here? (but I have not fully digested the changes)

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7082af/32134/summary.html
**COMMIT**: 662bfca19feb7515fbbd62653e32c1c0feaae688
**CMSSW**: CMSSW_13_1_X_2023-04-25-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/41361/32134/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially removed 1 lines from the logs
 - Reco comparison results: 11 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3460685
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3460657
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-04-25-1100+7082af/56664/triggerResults): no differences found

vlimant commented:
+1

civanch commented:
+1

emanueleusai commented:
+1

mandrenguyen commented:
+1

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41361/35498

- This PR adds an extra 92KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/ParameterSet/python/Config.py  modified in PR(s): #37996

cmsbuild commented:
Pull request #41361 was updated. @smuzaffar, @civanch, @Dr15Jones, @swertz, @vlimant, @makortel, @micsucmed, @emanueleusai, @mdhildreth, @cmsbuild, @clacaputo, @syuvivida, @simonepigazzini, @pmandrik, @mandrenguyen, @rvenditti can you please check and sign again.

emanueleusai commented:
please test

Dr15Jones commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41361/35502

- This PR adds an extra 72KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/ParameterSet/python/Config.py  modified in PR(s): #37996

cmsbuild commented:
Pull request #41361 was updated. @smuzaffar, @civanch, @Dr15Jones, @swertz, @vlimant, @makortel, @micsucmed, @emanueleusai, @mdhildreth, @clacaputo, @syuvivida, @simonepigazzini, @pmandrik, @mandrenguyen, @rvenditti can you please check and sign again.

cmsbuild commented:
-1

**Failed Tests**: UnitTests RelVals-INPUT 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7082af/32535/summary.html
**COMMIT**: e081f54665aec442407075778edebe8c7cd39e8d
**CMSSW**: CMSSW_13_2_X_2023-05-10-1200/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/41361/32535/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found errors in the following unit tests:

<pre>---> test condTestRegression had ERRORS
</pre>

## RelVals-INPUT

The relvals timed out after 4 hours.


## Comparison Summary



**Summary**:
 - You potentially removed 9 lines from the logs
 - Reco comparison results: 3 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3460836
 - DQMHistoTests: Total failures: 0
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3460814
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_2_X_2023-05-10-1200+7082af/56984/triggerResults): no differences found

makortel commented:
@cmsbuild, please test

The unit test fails already in IBs, but let's see if another round helps with the RelVals-INPUT

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7082af/32549/summary.html
**COMMIT**: e081f54665aec442407075778edebe8c7cd39e8d
**CMSSW**: CMSSW_13_2_X_2023-05-10-1200/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/41361/32549/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially removed 3 lines from the logs
 - Reco comparison results: 6 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3460836
 - DQMHistoTests: Total failures: 9
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3460805
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_2_X_2023-05-10-1200+7082af/56995/triggerResults): no differences found

emanueleusai commented:
+1

makortel commented:
+1

civanch commented:
+1

vlimant commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

rappoccio commented:
+1
